truffle/src/main/ruby/core/rubinius/bootstrap/thread.rb

```diff
@@ -41,4 +41,36 @@ def self.kill(thread)
     thread.kill
   end
 
+  def raise(exc=undefined, msg=nil, trace=nil)
+    return self unless alive?
+
+    if undefined.equal? exc
+      no_argument = true
+      exc         = nil
+    end
+
+    if exc.respond_to? :exception
+      exc = exc.exception msg
+      Kernel.raise TypeError, 'exception class/object expected' unless Exception === exc
+      exc.set_backtrace trace if trace
+    elsif no_argument
+      exc = RuntimeError.exception nil
+    elsif exc.kind_of? String
+      exc = RuntimeError.exception exc
+    else
+      Kernel.raise TypeError, 'exception class/object expected'
+    end
+
+    if $DEBUG
+      STDERR.puts "Exception: #{exc.message} (#{exc.class})"
+    end
+
+    if self == Thread.current
+      Kernel.raise exc
+    else
+      exc.capture_backtrace! 2 unless exc.backtrace?
+      Truffle::Primitive.thread_raise self, exc
+    end
+  end
+
```

Comment:
Where does this code come from? If it is modified it should be in api/shims.
If we can work with the original code, it is usually better, so in this case we could implement the `thread_raise` Rubinius primitive, right? (In ThreadPrimitiveNodes) See `VMPrimitiveNodes` for an example of Rubinius primitives nodes.